File: xamarin/Samples/AndorraTelecom-iOS/ViewControllers/MenuViewController.cs
using System;
using UIKit;
using LiferayScreens;
using Foundation;
using AndorraTelecomiOS.Util;
using CoreGraphics;

namespace AndorraTelecomiOS
{
    public partial class MenuViewController : UIViewController, IWebScreenletDelegate, ICallMeBackDelegate
    {
        protected MenuViewController(IntPtr handle) : base(handle) { }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();

            AddLogoToNavigationItem();

            AddMenuChangeLanguage();

            LoadWebScreenlet();

            LoadCallMeBackPopOver();
        }

        public void AddLogoToNavigationItem()
        {
            UIImageView ImageView = NavItem.AddLogo();
            NavigationItem.TitleView = ImageView;
        }

        public void AddMenuChangeLanguage()
        {
            var RightButton = NavItem.AddButtonChangeLanguage(LanguageHelper.ThreeLanguageLetters, this);
            NavigationItem.SetRightBarButtonItem(RightButton, true);

            this.viewCallMeBack.Hidden = true;
        }

        public void LoadWebScreenlet()
        {
            var Url = LanguageHelper.Url(LanguageHelper.Pages.Index);

            var config = new WebScreenletConfigurationBuilder(Url)
                .SetWithWebType(WebType.Other)
                .AddJsWithLocalFile("js/menu_js")
                .AddCssWithLocalFile("css/menu_css")
                .Load();

            webScreenlet.ThemeName = "andorra";
            webScreenlet.Configuration = config;
            webScreenlet.IsScrollEnabled = false;
            webScreenlet.BackgroundColor = Colors.LightPurple;
            webScreenlet.Delegate = this;
            webScreenlet.Load();
        }

        public void LoadCallMeBackPopOver()
        {
            CallMeBackView.Delegate = this; 

            var CmbView = CallMeBackView.Create();
            CmbView.Frame = callMeBackView.Bounds;
            callMeBackView.AddSubview(CmbView);
        }

        /* IWebScreenletDelegate */

        [Export("onWebLoad:url:")]
        public virtual void OnWebLoad(WebScreenlet screenlet, string url)
        {
            Console.WriteLine($"WebScreenlet URL display successfully: {url}");
            AttachClickHeaderCallMeBack();
        }

        [Export("screenlet:onError:")]
        public virtual void Screenlet(WebScreenlet screenlet, NSError error)
        {
            Console.WriteLine($"WebScreenlet URL display failed: {error.DebugDescription}");
        }

        [Export("screenlet:onScriptMessageNamespace:onScriptMessage:")]
        public virtual void Screenlet(WebScreenlet screenlet, string namespace_, string Message)
        {
            Console.WriteLine($"WebScreenlet onScriptMessage -> namespace: {namespace_}, message: {Message}");

            switch(namespace_)
            {
                case "call-me-back":
                    Console.WriteLine("Call me back popover");
                    CreatePopOverCallMeBack(Message);
                    break;
                case "click-button":
                    Console.WriteLine("Go to next forfet");
                    GoNextForfet(Message);
                    break;
                case "map":
                    Console.WriteLine("Go to map");
                    GoMap();
                    break;
                default:
                    Console.WriteLine("Invalid event");
                    break;
            }
        }

        /* UIViewController */

        public override void PrepareForSegue(UIStoryboardSegue segue, NSObject sender)
        {
            if(segue.Identifier == "forfet")
            {
                var LanguageBundle = NSBundle.FromPath(NSBundle.MainBundle.PathForResource(LanguageHelper.Language, "lproj"));
                var BackItem = new UIBarButtonItem();
                BackItem.Title = LanguageBundle.LocalizedString("Back", null);
                BackItem.TintColor = UIColor.White;
                NavigationItem.BackBarButtonItem = BackItem;

                var ViewController = segue.DestinationViewController as ForfetViewController;

                if(ViewController != null)
                {
                    switch(sender.ToString())
                    {
                        case "0":
                            ViewController.ForfetUrl = LanguageHelper.Pages.Mobile;
                            break;
                        case "1":
                            ViewController.ForfetUrl = LanguageHelper.Pages.Roaming;
                            break;
                        case "2":
                            ViewController.ForfetUrl = LanguageHelper.Pages.Paquete69;
                            break;
                        case "3":
                            ViewController.ForfetUrl = LanguageHelper.Pages.Optima;
                            break;
                        default:
                            Console.WriteLine("Invalid page");
                            break;
                    }
                }
            }

        }

        /* ICallMeBackDelegate */

        public void ShowLegalConditions(CallMeBackView CallMeBackView)
        {
            Console.WriteLine("Show Legal Conditions");

            var AlertController = new UIAlertController();
            AlertController.Title = "\n\n\n\n\n\n\n\n\n\n\n\n\n";
            AlertController.ModalPresentationStyle = UIModalPresentationStyle.PageSheet;
            AlertController.View.Layer.CornerRadius = 15;
            AlertController.View.TintColor = Colors.DarkPurple;
            var Subview = AlertController.View.Subviews[0].Subviews[0] as UIView;
            foreach (var V in Subview.Subviews)
            {
                V.BackgroundColor = Colors.Pink;
            }

            var LegalView = LegalConditionsView.Create();

            var Margin = 5;
            var Rect = new CGRect(Margin, Margin, AlertController.View.Bounds.Size.Width - Margin * 2, AlertController.View.Bounds.Size.Height - Margin * 26);
            LegalView.Frame = Rect;
            AlertController.View.AddSubview(LegalView);

            var LanguageBundle = RetrieveLanguageBundle(LanguageHelper.Language);

            var AlertActionAccept = UIAlertAction.Create(LanguageBundle.LocalizedString("Accept", null), UIAlertActionStyle.Default, (obj) => this.callMeBackView.LegalConditionsChange(CallMeBackView, true));
            var AlertActionCancel = UIAlertAction.Create(LanguageBundle.LocalizedString("Cancel", null), UIAlertActionStyle.Cancel, (obj) => this.callMeBackView.LegalConditionsChange(CallMeBackView, false));
            AlertController.AddAction(AlertActionAccept);
            AlertController.AddAction(AlertActionCancel);
            this.PresentViewController(AlertController, true, null);
        }

        public void ShowAlertLegalNotAccepted(CallMeBackView CallMeBackView, string Title, string Message)
        {
            Console.WriteLine("Show Alert Legal Not Accepted");
        }

        /* Private methods */

        void GoNextForfet(string Message)
        {
            PerformSegue("forfet", NSObject.FromObject(Message));
        }

        void GoMap()
        {
            PerformSegue("map", null);
        }

        async void CreatePopOverCallMeBack(string Message)
        {
            this.labelCallMeBack.Text = Message;
            await UIView.AnimateAsync(0, () => AnimateCallMeBack());
        }

        void AnimateCallMeBack()
        {
            this.heightCallMeBack.Constant = 50;
            this.viewCallMeBack.Superview.LayoutIfNeeded();
            this.viewCallMeBack.Hidden = false;
        }

        void AttachClickHeaderCallMeBack()
        {
            Action CallMeBackActions = () =>
            {
                UIView.Animate(0.5, () =>
                {
                    var bigSize = 400;
                    var smallSize = 50;

                    if (this.heightCallMeBack.Constant.Equals(bigSize))
                    {
                        this.heightCallMeBack.Constant = smallSize;
                        this.viewCallMeBack.Superview.LayoutIfNeeded();
                    }
                    else
                    {
                        this.heightCallMeBack.Constant = bigSize;
                        this.viewCallMeBack.Superview.LayoutIfNeeded();
                    }
                });
            };

            var headerGesture = new UITapGestureRecognizer(CallMeBackActions);
            this.headerCallMeBack.AddGestureRecognizer(headerGesture);
        }

        static NSBundle RetrieveLanguageBundle(string Language)
        {
            var Path = NSBundle.MainBundle.PathForResource(Language, "lproj");
            return NSBundle.FromPath(Path);
        }
    }
}

File: ios/Samples/AndorraTelecomPublic/AndorraTelecomPublic/ViewControllers/MenuViewController.swift
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import UIKit
import LiferayScreens
import Hokusai

@objcMembers
class MenuViewController: UIViewController, WebScreenletDelegate, CallMeBackDelegate {

    @IBOutlet weak var webScreenlet: WebScreenlet!
    @IBOutlet weak var heightCallMeBack: NSLayoutConstraint!
    @IBOutlet weak var headerCallmeBack: UIView!
    @IBOutlet weak var viewCallmeBack: UIView!
    @IBOutlet weak var labelCallMeBack: UILabel!
    @IBOutlet weak var callMeBack: CallMeBackView!
    @IBOutlet weak var coverForClick: UIVisualEffectView!

    var coverNavegationBar: UIView?

    override func viewDidLoad() {
        super.viewDidLoad()

        addLogoToNavigationBar()
        modifyHeightCallMeBack(height: 0)
        callMeBack.delegate = self

        buttonChangeLanguage(language: LanguageHelper.shared().threeLettersFormatted)
        textButtonBack()
        loadWebScreenlet()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    func modifyHeightCallMeBack(height: CGFloat) {
        heightCallMeBack.constant = height
        self.viewCallmeBack.layoutIfNeeded()
    }

    func addLogoToNavigationBar() {
        let logo = UIImage(named: "Logo") as UIImage?
        let imageView = UIImageView(image:logo)
        imageView.frame.size.width = 100
        imageView.frame.size.height = 32
        imageView.contentMode = UIViewContentMode.scaleAspectFill

        self.navigationItem.titleView = imageView
    }

    func buttonChangeLanguage(language: String) {
        navigationItem.rightBarButtonItem = UIBarButtonItem(
            title: language,
            style: .plain,
            target: self,
            action: Colorful.flag ? #selector(menuChangeLanguageColorful) : #selector(menuChangeLanguage))
    }

    func menuChangeLanguage() {
        let actionSheetController: UIAlertController = UIAlertController(title: "language".localized(), message: "choose".localized(), preferredStyle: .actionSheet)

        for value in LanguageHelper.shared().listLanguages {
            let itemAction: UIAlertAction = UIAlertAction(title: value, style: .default) { _ -> Void in
                self.actionChangeLanguage(value: value)
            }
            actionSheetController.addAction(itemAction)
        }

        let cancelAction: UIAlertAction = UIAlertAction(title: "cancel".localized(), style: .cancel) { _ -> Void in
            actionSheetController.dismiss(animated: true, completion: nil)
        }

        actionSheetController.addAction(cancelAction)
        self.present(actionSheetController, animated: true, completion: nil)
    }

    func menuChangeLanguageColorful() {
        let hokusai = Hokusai()

        hokusai.colors = HOKColors(
            backGroundColor: UIColor.lightPurple,
            buttonColor: UIColor.darkPurple,
            cancelButtonColor: UIColor.darkPurple,
            fontColor: UIColor.white
        )

        for value in LanguageHelper.shared().listLanguages {
            hokusai.addButton(value) {
                self.actionChangeLanguage(value: value)
            }
        }
        hokusai.show()
    }

    func actionChangeLanguage(value: String) {
        LanguageHelper.shared().change(language: value)
        self.buttonChangeLanguage(language: LanguageHelper.shared().threeLettersFormatted)
        self.loadWebScreenlet()
        self.textButtonBack()
        self.callMeBack.setTextOutlets()
    }

    func textButtonBack() {
        let backItem = UIBarButtonItem(title: "back".localized(), style: .plain, target: nil, action: nil)
        navigationItem.backBarButtonItem = backItem
    }

    func attachClickHeaderCallBack() {
        let headerGesture = UITapGestureRecognizer(target: self, action:  #selector (MenuViewController.callBackActions(sender:)))

        let touchUpViewGesture = UITapGestureRecognizer(target: self, action:  #selector (MenuViewController.callBackActions(sender:)))

        let navBarViewGesture = UITapGestureRecognizer(target: self, action:  #selector (MenuViewController.callBackActions(sender:)))

        self.headerCallmeBack.addGestureRecognizer(headerGesture)
        self.coverForClick.addGestureRecognizer(touchUpViewGesture)
        self.coverNavegationBar!.addGestureRecognizer(navBarViewGesture)

    }

    func callBackActions(sender: UITapGestureRecognizer) {
        UIView.animate(withDuration: 0.5, animations: {
            let bigSize: CGFloat = 400.0
            let smallSize: CGFloat = 50.0

            if self.heightCallMeBack.constant.isEqual(to: bigSize) {
                self.heightCallMeBack.constant = smallSize
                self.viewCallmeBack.superview?.layoutIfNeeded()

                self.navigationController?.navigationBar.dismissChangeBackgroundColor(view: self.coverNavegationBar!, isUserInteraccionEnable: false)
                self.coverForClick.dismissChangeBackgroundColor(isHidden: true)
                self.coverNavegationBar?.isUserInteractionEnabled = false
            }
            else {
                self.heightCallMeBack.constant = bigSize
                self.viewCallmeBack.superview?.layoutIfNeeded()

                self.navigationController?.navigationBar.animatedChangeBackgroundColor(view: self.coverNavegationBar!, isUserInteraccionEnable: true)
                self.coverForClick.animatedChangeBackgroundColor(isHidden: false)
                self.coverNavegationBar?.isUserInteractionEnabled = true
            }

        })
    }

    func prepareCoverNavegationBar() {
        let xposition = self.navigationController?.navigationBar.frame.origin.x
        let yposition = (self.navigationController?.navigationBar.frame.origin.y)! - 60
        let height = (self.navigationController?.navigationBar.frame.size.height)! + 40
        let width = self.navigationController?.navigationBar.frame.size.width

        let frame = CGRect(x: xposition!, y: yposition, width: width!, height: height)

        self.coverNavegationBar = UIView(frame: frame)
        self.navigationController?.navigationBar.addSubview(self.coverNavegationBar!)
        self.coverNavegationBar?.isUserInteractionEnabled = false
    }

    func loadWebScreenlet() {
        let webScreenletConfiguration = WebScreenletConfigurationBuilder(url: LanguageHelper.shared().url(page: .index))
            .set(webType: .other)
            .addCss(localFile: "menu")
            .addJs(localFile: "menu")
            .load()
        webScreenlet.configuration = webScreenletConfiguration
        webScreenlet.isScrollEnabled = false
        webScreenlet.backgroundColor = UIColor.lightPurple

        webScreenlet.load()
        webScreenlet.delegate = self
    }

    func onWebLoad(_ screenlet: WebScreenlet, url: String) {
        prepareCoverNavegationBar()
        attachClickHeaderCallBack()
    }

    func screenlet(_ screenlet: WebScreenlet, onScriptMessageNamespace namespace: String, onScriptMessage message: String) {
        switch namespace {
            case "call-me-back":
                createPopOverCallMeBack(message: message)
            case "click-button":
                goNextForfet(position: message)
            case "map":
                goToMap()
            default:
                print("Any event")
        }

    }

    func createPopOverCallMeBack(message: String) {
        labelCallMeBack.text = message
        UIView.animate(withDuration: 1, animations: {
            self.heightCallMeBack.constant = 50
            self.viewCallmeBack.superview?.layoutIfNeeded()
        })
    }

    func goNextForfet(position: String) {
        performSegue(withIdentifier: "forfet", sender: position)
    }

    func goToMap() {
        performSegue(withIdentifier: "map", sender: nil)
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "forfet" ,
            let vc = segue.destination as? ForfetViewController {
            switch sender as! String {
            case "0":
                vc.url = LanguageHelper.shared().url(page: .mobile)
            case "1":
                vc.url = LanguageHelper.shared().url(page: .roaming)
            case "2":
                vc.url = LanguageHelper.shared().url(page: .paquete69)
            case "3":
                vc.url = LanguageHelper.shared().url(page: .optima)
            default:
                print("Sorry, we don't have more cases")
            }

        }
    }

    func showAlertLegalNotAccepted(callMeBackView: CallMeBackView, title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: UIAlertControllerStyle.alert)
        alert.addAction(UIAlertAction(title: "Ok", style: UIAlertActionStyle.default, handler: nil))
        alert.colorfull()
        self.present(alert, animated: true, completion: nil)
    }

    func showLegalCoditions(callMeBackView: CallMeBackView) {
        let alertController = UIAlertController(title: "\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n", message: nil, preferredStyle: UIAlertControllerStyle.actionSheet)

        let margin: CGFloat = 10.0
        let rect = CGRect(x: margin, y: margin, width: alertController.view.bounds.size.width - margin * 6.0, height: 340)

        let legalConditionView = LegalConditionsView(frame: rect)

        alertController.view.addSubview(legalConditionView)

        let acceptAction = UIAlertAction(title: "accept".localized(), style: .default, handler: { (_: UIAlertAction!) in self.callMeBack.legalConditionsChange(isAccepted: true) })

        let cancelAction = UIAlertAction(title: "cancel".localized(), style: .cancel, handler: { (_: UIAlertAction!) in self.callMeBack.legalConditionsChange(isAccepted: false) })

        alertController.addAction(acceptAction)
        alertController.addAction(cancelAction)

        alertController.colorfull()

        DispatchQueue.main.async {
            self.present(alertController, animated: true, completion: {})
        }

    }
}
